NArray/Matrix as const throws compiler error

Hey,
in my simplified example I want to call a function with a const refernce to a NArray/Matrix.
But then I get an compiler error.
~~~shell
main.cpp:27:30: error: passing ‘const local_type {aka const dash::LocalMatrixRef<unsigned int, 2, 2, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >}’ as ‘this’ argument discards qualifiers [-fpermissive]
~~~
I made a very reduced example in [feat-283-constMatrix](https://github.com/dash-project/dash/blob/feat-283-constMatrix/dash/examples/ex.12.feat-283/main.cpp).
If the parameter in [line12](https://github.com/dash-project/dash/blob/feat-283-constMatrix/dash/examples/ex.12.feat-283/main.cpp#L12) is const then the compiler error arises.

This example is without really meaning, but to show the problem. 
The real example would be [cowichan_product](https://github.com/dash-project/dash-apps/blob/master/cowichan/product/main.cpp).
Everything is fine as long as NArray/Matrix is not a const parameter.

With the latest changes my examples from above don't comile with const parameter anymore.
Trying to compile the example from the feat branch leads to following error.
[constMatrix compiler Error.txt](https://github.com/dash-project/dash/files/787597/constMatrix.compiler.Error.txt)
An Extract:

~~~c++
/home/dash/opt/dash-0.3.0/include/dash/matrix/internal/LocalMatrixRef-inl.h:442:21: error:
cannot convert 
‘dash::Matrix<      unsigned int, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >* const’
to
‘dash::Matrix<const unsigned int, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >*’
in assignment
   ref._refview._mat = _refview._mat;
                     ^
~~~

There is another compiler error concerning `const matrices`.
In cowichan/randmat there's a print2D function which prints matrices to sdt::out.
The old signature looks like this: 
~~~C++
void print2d( T& mat )
~~~
and the new one like this:
~~~C++
void print2d( const T& mat )
~~~

with this i get the following compiler error:
~~~C++
randmat.cpp: In instantiation of ‘void print2d(const T&) [with T = dash::Matrix<unsigned char, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >]’:
randmat.cpp:79:30:   required from here
randmat.cpp:12:34: error: no match for call to ‘(const dash::Matrix<unsigned char, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >) (int&, int&)’
       cout << std::setw(3) << mat(i,j)<< " ";
                                  ^
In file included from /home/lehmannb/opt/dash-0.3.0/include/dash/Matrix.h:645:0,
                 from /home/lehmannb/opt/dash-0.3.0/include/dash/Container.h:72,
                 from /home/lehmannb/opt/dash-0.3.0/include/libdash.h:39,
                 from randmat.cpp:1:
/home/lehmannb/opt/dash-0.3.0/include/dash/matrix/internal/Matrix-inl.h:488:1: note: candidate: dash::Matrix<T, NumDimensions, IndexT, PatternT>::reference dash::Matrix<T, NumDimensions, IndexT, PatternT>::operator()(Args ...) [with Args = {int, int}; ElementT = unsigned char; int NumDimensions = 2; IndexT = long int; PatternT = dash::BlockPattern<2, (dash::MemArrange)1u, long int>; dash::Matrix<T, NumDimensions, IndexT, PatternT>::reference = dash::GlobRef<unsigned char>] <near match>
 Matrix<T, NumDim, IndexT, PatternT>
 ^
/home/lehmannb/opt/dash-0.3.0/include/dash/matrix/internal/Matrix-inl.h:488:1: note:   passing ‘const dash::Matrix<unsigned char, 2, long int, dash::BlockPattern<2, (dash::MemArrange)1u, long int> >*’ as ‘this’ argument discards qualifiers
make: *** [randmat] Error 1
~~~
[Here](https://github.com/dash-project/dash-apps/blob/6dc88bbf842a1484469cbd8dd993f542f39ac7ef/cowichan/randmat/randmat.cpp#L9) is the code i am talking about.
If you want to try it yourself, please be sure to use this commit i linked to.

--- a/dash/include/dash/algorithm/MinMax.h
+++ b/dash/include/dash/algorithm/MinMax.h
@@ -53,6 +53,7 @@ const ElementType * min_element(
   Compare             compare
     = std::less<const ElementType &>())
 {
+  typedef typename std::decay<ElementType>::type      value_t;
 #ifdef DASH_ENABLE_OPENMP
   dash::util::UnitLocality uloc;
   auto n_threads = uloc.num_domain_threads();
@@ -65,7 +66,7 @@ const ElementType * min_element(
     int           min_idx_l  = 0;
     ElementType   min_val_l  = *l_range_begin;
 
-    typedef struct min_pos_t { ElementType val; size_t idx; } min_pos;
+    typedef struct min_pos_t { value_t val; size_t idx; } min_pos;
 
     DASH_LOG_DEBUG("dash::min_element", "local range size:", l_size);
     int       align_bytes      = uloc.cache_line_size(0);
@@ -105,8 +106,6 @@ const ElementType * min_element(
       // Cannot use explicit private(min_val_t) as ElementType might
       // not be default-constructible:
       #pragma omp for schedule(static)
-//    #pragma ivdep
-//    #pragma vector aligned nontemporal
       for (int i = 0; i < l_size; i++) {
         const ElementType & val_t = *(l_range_begin + i);
         if (compare(val_t, min_vals_t[t_id].val)) {
@@ -164,6 +163,7 @@ GlobIter<ElementType, PatternType> min_element(
   typedef dash::GlobIter<ElementType, PatternType> globiter_t;
   typedef PatternType                               pattern_t;
   typedef typename pattern_t::index_type              index_t;
+  typedef typename std::decay<ElementType>::type      value_t;
 
   // return last for empty array
   if (first == last) {
@@ -219,8 +219,8 @@ GlobIter<ElementType, PatternType> min_element(
   trace.exit_state("barrier");
 
   typedef struct {
-    ElementType value;
-    index_t     g_index;
+    value_t  value;
+    index_t  g_index;
   } local_min_t;
 
   std::vector<local_min_t> local_min_values(team.size());
@@ -352,7 +352,7 @@ GlobIter<ElementType, PatternType> max_element(
  */
 template <
   class ElementType,
-  class Compare = std::greater<const ElementType &> >
+  class Compare = std::greater<ElementType &> >
 const ElementType * max_element(
   /// Iterator to the initial position in the sequence
   const ElementType * first,
@@ -360,7 +360,7 @@ const ElementType * max_element(
   const ElementType * last,
   /// Element comparison function, defaults to std::less
   Compare             compare
-    = std::greater<const ElementType &>())
+    = std::greater<ElementType &>())
 {
   // Same as min_element with different compare function
   return dash::min_element(first, last, compare);

--- a/dash/include/dash/matrix/MatrixRef.h
+++ b/dash/include/dash/matrix/MatrixRef.h
@@ -317,6 +317,31 @@ public:
    * \returns  A global reference to the element at the given global
    *           coordinates.
    */
+  template<typename ... Args>
+  const_reference at(
+    /// Global coordinates
+    Args... args) const;
+
+  /**
+   * Fortran-style subscript operator.
+   * As an example, the operation \c matrix(i,j) is equivalent to
+   * \c matrix[i][j].
+   *
+   * \returns  A global reference to the element at the given global
+   *           coordinates.
+   */
+  const_reference at(
+    /// Global coordinates
+    const ::std::array<index_type, NumDimensions> & coords) const;
+
+  /**
+   * Fortran-style subscript operator.
+   * As an example, the operation \c matrix(i,j) is equivalent to
+   * \c matrix[i][j].
+   *
+   * \returns  A global reference to the element at the given global
+   *           coordinates.
+   */
   reference at(
     /// Global coordinates
     const ::std::array<index_type, NumDimensions> & coords);

--- a/dash/include/dash/matrix/internal/MatrixRef-inl.h
+++ b/dash/include/dash/matrix/internal/MatrixRef-inl.h
@@ -447,19 +447,32 @@ MatrixRef<T, NumDim, CUR, PatternT>
 
 template <typename T, dim_t NumDim, dim_t CUR, class PatternT>
 template <typename ... Args>
+inline typename MatrixRef<T, NumDim, CUR, PatternT>::const_reference
+MatrixRef<T, NumDim, CUR, PatternT>
+::at(Args... args) const
+{
+  return at(std::array<index_type, NumDim> {{
+              static_cast<index_type>(args)...
+            }});
+}
+
+template <typename T, dim_t NumDim, dim_t CUR, class PatternT>
+template <typename ... Args>
 inline typename MatrixRef<T, NumDim, CUR, PatternT>::reference
 MatrixRef<T, NumDim, CUR, PatternT>
 ::at(Args... args)
 {
-  if(sizeof...(Args) != (NumDim - _refview._dim)) {
-    DASH_THROW(
-      dash::exception::InvalidArgument,
-      "MatrixRef.at(): Invalid number of arguments " <<
-      "expected " << (NumDim - _refview._dim) << " " <<
-      "got " << sizeof...(Args));
-  }
-  ::std::array<index_type, NumDim> coords = {{ static_cast<index_type>(args)... }};
-  return at(coords);
+  return at(std::array<index_type, NumDim> {{
+              static_cast<index_type>(args)...
+            }});
+}
+
+template <typename T, dim_t NumDim, dim_t CUR, class PatternT>
+inline typename MatrixRef<T, NumDim, CUR, PatternT>::const_reference
+MatrixRef<T, NumDim, CUR, PatternT>
+::at(const ::std::array<typename PatternT::index_type, NumDim> & coords) const
+{
+  return _refview.global_reference(coords);
 }
 
 template <typename T, dim_t NumDim, dim_t CUR, class PatternT>

--- a/dash/test/algorithm/MinElementTest.cc
+++ b/dash/test/algorithm/MinElementTest.cc
@@ -27,9 +27,11 @@ TEST_F(MinElementTest, TestFindArrayDefault)
   LOG_MESSAGE("Waiting for other units to initialize array values");
   array.barrier();
   LOG_MESSAGE("Finished initialization of array values");
-  // Run min_element on complete array
-  auto found_gptr = dash::min_element(array.begin(),
-                                      array.end());
+  // Run min_element on complete array, use const reference to test
+  // const-correctness:
+  const Array_t & array_cref = array;
+  auto found_gptr = dash::min_element(array_cref.begin(),
+                                      array_cref.end());
   // Check that a minimum has been found (found != last):
   LOG_MESSAGE("Completed dash::min_element");
   // Run min_element on complete array

--- a/dash/test/container/MatrixTest.cc
+++ b/dash/test/container/MatrixTest.cc
@@ -1388,6 +1388,24 @@ ValueT global_sum_rows(const int       nelts,
   return global_sum;
 }
 
+template <
+  class MatrixT,
+  class ValueT = typename MatrixT::value_type >
+ValueT global_sum_elems(const int       nelts,
+                        const MatrixT & matIn)
+{
+  auto   glbRows    = matIn.pattern().extents()[0];
+  auto   glbCols    = matIn.pattern().extents()[1];
+  ValueT global_sum = 0;
+
+  for (int i = 0; i < glbRows; ++i) {
+    for (int j = 0; j < glbCols; ++j) {
+      global_sum += matIn(i,j);
+    }
+  }
+  return global_sum;
+}
+
 TEST_F(MatrixTest, ConstMatrixRefs)
 {
   using value_t = unsigned int;
@@ -1408,23 +1426,25 @@ TEST_F(MatrixTest, ConstMatrixRefs)
   }
   dash::barrier();
 
-  auto local_rows_sum = local_sum_rows(nelts, mat);
-  auto local_elem_sum = std::accumulate(
-                          mat.lbegin(),
-                          mat.lend(),
-                          0, std::plus<value_t>());
+  auto local_rows_sum  = local_sum_rows(nelts, mat);
+  auto local_range_sum = std::accumulate(
+                           mat.lbegin(),
+                           mat.lend(),
+                           0, std::plus<value_t>());
 
-  EXPECT_EQ_U(local_elem_sum, local_rows_sum);
+  EXPECT_EQ_U(local_range_sum, local_rows_sum);
 
   dash::barrier();
 
-  auto global_rows_sum = global_sum_rows(nelts, mat);
-  auto global_elem_sum = std::accumulate(
-                           mat.begin(),
-                           mat.end(),
-                           0, std::plus<value_t>());
+  auto global_rows_sum  = global_sum_rows(nelts, mat);
+  auto global_elems_sum = global_sum_elems(nelts, mat);
+  auto global_range_sum = std::accumulate(
+                            mat.begin(),
+                            mat.end(),
+                            0, std::plus<value_t>());
 
-  EXPECT_EQ_U(global_elem_sum, global_rows_sum);
+  EXPECT_EQ_U(global_range_sum, global_rows_sum);
+  EXPECT_EQ_U(global_range_sum, global_elems_sum);
 }
 
 TEST_F(MatrixTest, SubViewMatrix3Dim)

